### ISSUE TITLE
IntegrateData() returns error in Seurat 4.0.1

### ISSUE THREAD

**Issue by nkomarov1410** [satijalab/seurat#4266]
Hi, 

I'm trying to re-run some analysis that was previously performed in version 3.0.0, which worked fine for a former group member. I cannot recreate their results because I'm running into an error while performing the IntegrateData function, the error is as follows: 

```
> organoid.combined <- IntegrateData(anchorset = organoid.anchors, dims = 1:20)
Merging dataset 1 into 2
Extracting anchors for merged samples
Finding integration vectors
Finding integration vector weights
Error in idx[i, ] <- res[[i]][[1]] : 
  number of items to replace is not a multiple of replacement length
```
Previous posts on a similar issue didn't provide an answer: 

https://stackoverflow.com/questions/38738347/why-do-i-get-number-of-items-to-replace-is-not-a-multiple-of-replacement-length

https://stat.ethz.ch/pipermail/r-help/2011-November/296083.html

https://github.com/satijalab/seurat/issues/3936

This post : https://www.gitmemory.com/issue/satijalab/seurat/3868/750938936 appears to have a similar problem (I also have a very small number of cells), but unfortunately there was never a reply. 

Please let me know if there is anything I can do!

**Comment by margoemont** [comment]
I had a similar issue when I switched from seurat 3 to 4, for me the issue was the introduction of the k.weight parameter in IntegrateData, which by default is set at 100, but should be less than the smallest number of cells you're integrating with. So, for example, if organoid.list contains objects with 100, 40, 200, and 300 cells, I would run
`organoid.combined <- IntegrateData(anchorset = organoid.anchors, dims = 1:20, k.weight = 39)`
I hope this helps!

**Comment by timoast** [comment]
@nkomarov1410 can you post the full code you're running? Also, are you trying to run this on a saved anchor object that was generated with Seurat v3 or are you computing a new anchor object?

**Comment by nkomarov1410** [comment]
Hi, 

@mem728 's advice made the code work, the k.weight was set to default, and reducing it to below the smallest number of cells solved the issue. 

I was computing a new anchor object

The full code up to this point is as follows: 

`min_gene_number <- 400
min.cells <- 1

#######
#Load and label

sample_1 <- "Sensory40M_Ham3_nofilt"
sample_2 <- "Sensory_1_mix_multiple_merge"
sample_3 <- "Sensory2single"

#######
###
sample_1_1.data <- read.table()
sample_2_1.data <- read.table()
sample_3_1.data <- read.table()

#####
# Create and setup Seurat objects for each dataset
#Sample_1
sample_1_1_seurat <- CreateSeuratObject(counts = sample_1_1.data, min.cells = min.cells, min.features = min_gene_number)
sample_1_seurat <- sample_1_1_seurat
sample_1_seurat$stim <- "sample_1"
sample_1_seurat <- NormalizeData(sample_1_seurat)
sample_1_seurat <- FindVariableFeatures(sample_1_seurat, do.plot = T, display.progress = F)
sample_1_seurat <- ScaleData(sample_1_seurat)
sample_1_seurat@meta.data$tech <- sample_1

#sample_2
sample_2_1_seurat <- CreateSeuratObject(counts = sample_2_1.data, min.cells = min.cells, min.features= min_gene_number)
sample_2_seurat <- sample_2_1_seurat
sample_2_seurat$stim <- "sample_2"
sample_2_seurat <- NormalizeData(sample_2_seurat)
sample_2_seurat <- FindVariableFeatures(sample_2_seurat, do.plot = T, display.progress = F)
sample_2_seurat <- ScaleData(sample_2_seurat)
sample_2_seurat@meta.data$tech <- sample_2

#sample_3
sample_3_1_seurat <- CreateSeuratObject(counts = sample_3_1.data, min.cells = min.cells, min.features = min_gene_number)
sample_3_seurat <- sample_3_1_seurat
sample_3_seurat$stim <- "sample_3"
sample_3_seurat <- NormalizeData(sample_3_seurat)
sample_3_seurat <- FindVariableFeatures(sample_3_seurat, do.plot = T, display.progress = F)
sample_3_seurat <- ScaleData(sample_3_seurat)
sample_3_seurat@meta.data$tech <- sample_3

######
#QC assessment
######
#Merge the seurat objects
organoid.merged <- merge(sample_1_seurat, sample_2_seurat)
organoid.merged <- merge(organoid.merged, sample_3_seurat)

#####
### Mito genes
mito.genes <- grep(pattern = "^mt:", x = rownames(organoid.merged[["RNA"]]@data), value = TRUE)
percent.mito <- colSums(as.matrix(organoid.merged[["RNA"]]@counts[mito.genes, ]))/colSums(as.matrix(organoid.merged[["RNA"]]@counts))*100
organoid.merged <- AddMetaData(object = organoid.merged, metadata = percent.mito, col.name = "percent.mito")

# Suppl.Fig.5B:
VlnPlot(object = organoid.merged, features = c("nFeature_RNA", "nCount_RNA","percent.mito"),
        ncol = 3)
FeatureScatter(object = organoid.merged, feature1 = "nCount_RNA", feature2 = "nFeature_RNA", pt.size = 1.5)

### filter
organoid.merged <- subset(organoid.merged,subset = percent.mito<10 & nCount_RNA < 50000)
VlnPlot(object = organoid.merged, features = c("nFeature_RNA", "nCount_RNA","percent.mito"),
        ncol = 3)

organoid.merged <- NormalizeData(organoid.merged)
organoid.merged <- FindVariableFeatures(organoid.merged, selection.method = "vst",do.plot = T, display.progress = F)
all.genes <- rownames(x = organoid.merged)
organoid.merged <- ScaleData(organoid.merged)


# Integrate data
ob.list <- SplitObject(organoid.merged, split.by = "stim")
ob.list

for (i in 1:length(ob.list)) {
  ob.list[[i]] <- NormalizeData(ob.list[[i]], verbose = FALSE)
  ob.list[[i]] <- FindVariableFeatures(
    ob.list[[i]], selection.method = "vst",
    nfeatures = 2000, verbose = FALSE
  )
}

ob.list <- lapply(X = ob.list, FUN = function(x) {
  x <- NormalizeData(x)
  x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)
})

k.filter <- min(200, min(sapply(ob.list, ncol)))
k.filter
organoid.anchors <- Seurat::FindIntegrationAnchors(ob.list, k.filter = k.filter)

organoid.combined <- IntegrateData(anchorset = organoid.anchors, dims = 1:20, k.weight=39)
organoid.combined

DefaultAssay(organoid.combined) <- "integrated"`

**Comment by timoast** [comment]
Ok I'll close this now then if it's solved

**Comment by jeremycfd** [comment]
Hello,

I have now encountered this same issue, but I have verified that the default k.weight (100) is not larger than the smallest number of cells in my list of seurat objects. However, reducing k.weight arbitrarily still resolved the issue. I'm wondering if there is a better way to diagnose the maximum k.weight that can be used, since it seems in my case to be independent of the number of cells in the seurat objects being integrated.

Thanks!
